evennia/typeclasses/admin.py

```diff
@@ -1,68 +1,325 @@
+        for obj in self.deleted_objects:
+            handler = get_handler(obj)
+            handler.remove(obj.attr_key, category=obj.attr_category)
+        for instance in instances:
+            handler = get_handler(instance)
+            strattr = True if instance.attr_strvalue else False
+            value = instance.attr_value or instance.attr_strvalue
+            try:
+                handler.add(instance.attr_key, value, category=instance.attr_category, strattr=strattr,
+                            lockstring=instance.attr_lockstring)
+            except (TypeError, ValueError):
+                # catch errors in nick templates and continue
+                traceback.print_exc()
+                continue
+
 
+class AttributeInline(admin.TabularInline):
+    """
+    A handler for inline Tags. This class should be subclassed in the admin of your models,
+    and the 'model' and 'related_field' class attributes must be set. model should be the
+    through model (ObjectDB_db_tag', for example), while related field should be the name
+    of the field on that through model which points to the model being used: 'objectdb',
+    'msg', 'playerdb', etc.
     """
     # Set this to the through model of your desired M2M when subclassing.
     model = None
-    extra = 1
-    #form = AttributeForm
-    fields = ('attribute', 'key', 'value', 'strvalue')
+    form = AttributeForm
+    formset = AttributeFormSet
+    related_field = None  # Must be 'objectdb', 'playerdb', 'msg', etc. Set when subclassing
     raw_id_fields = ('attribute',)
-    readonly_fields = ('key', 'value', 'strvalue')
+    readonly_fields = ('attribute',)
+    extra = 0
 
-    def key(self, instance):
-        if not instance.id:
-            return "Not yet set or saved."
-        return '<strong><a href="%s">%s</a></strong>' % (
-            reverse("admin:typeclasses_attribute_change",
-                    args=[instance.attribute.id]),
-            instance.attribute.db_key)
+    def get_formset(self, request, obj=None, **kwargs):
+        """
+        get_formset has to return a class, but we need to make the class that we return
+        know about the related_field that we'll use. Returning the class itself rather than
+        a proxy isn't threadsafe, since it'd be the base class and would change if multiple
+        people used the admin at the same time
+        """
+        formset = super(AttributeInline, self).get_formset(request, obj, **kwargs)
 
-    key.allow_tags = True
+        class ProxyFormset(formset):
+            pass
 
-    def value(self, instance):
-        if not instance.id:
-            return "Not yet set or saved."
-        return instance.attribute.value
+        ProxyFormset.related_field = self.related_field
+        return ProxyFormset
 
-    def strvalue(self, instance):
-        if not instance.id:
-            return "Not yet set or saved."
-        return instance.attribute.strvalue
+# class AttributeInline(admin.TabularInline):
```

Comment:
Probably ok to remove this commented-out section; Attributes never appear away from an object.